### PR TITLE
rgw: need to 'open_object_section' before dump stats in 'RGWGetUsage_…

### PR DESCRIPTION
**Pull request by weiqiaomiao**
…ObjStore_S3::send_response()' function

Signed-off-by: weiqiaomiao <wei.qiaomiao@zte.com.cn>

### CODE REVIEW

**Comment by yehudasa** [comment]
@weiqiaomiao this is a public api (albeit rgw extension). Need to bring this change to the mailing list, people might be using it as is now.

**Comment by yehudasa** [comment]
@chenji-kael please take a look at this one

**Comment by chenji-kael** [comment]
@yehudasa @weiqiaomiao 
Thanks for your remind， I think it's just about json decode format issue, and does not much matter, and actually we use this api for production environment ,and change this may cause problems when we update ceph.

**Comment by weiqiaomiao** [comment]
@yehudasa I send a email to the mailing list（the email titile is 'RGW:add description  for GET usage admin api response entity 'stats entry''), and there is no one reponse it. I think  there is no people using this api except chenji. 
@chenji-kael One of our customers planning use this api, and  it is comfused to them what's the figure mean when they decode the api reponse data, add a description is a simple but easy way to help them quickly understand it.How about to fix this when you convenience?

**Comment by chenji-kael** [comment]
@weiqiaomiao 
please see this pr , maybe you need it
https://github.com/ceph/ceph/pull/8043

**Comment by yehudasa** [comment]
@weiqiaomiao @chenji-kael I suggest that we switch to the new api, but make it configurable so that current users can fall back to the old api. What do you think?

**Comment by weiqiaomiao** [comment]
@yehudasa  @chenji-kael  It is a good suggestion to me and  update the commit yet.

File: src/common/config_opts.h [hunk 1]
```diff
@@ -1433,6 +1433,8 @@ OPTION(rgw_swift_versioning_enabled, OPT_BOOL, false) // whether swift object ve
 
 OPTION(rgw_list_bucket_min_readahead, OPT_INT, 1000) // minimum number of entries to read from rados for bucket listing
 
+OPTION(rgw_getusage_dump_description, OPT_BOOL, true) // dump description of total stats for s3 GetUsage API
```
Comment (yehudasa) [inline src/common/config_opts.h]: @weiqiaomiao let's rename it to rgw_rest_getusage_op_compat, and default it to false (switch the flag's meaning)

File: src/rgw/rgw_rest_s3.cc [hunk 2]
```diff
@@ -429,10 +429,20 @@ void RGWGetUsage_ObjStore_S3::send_response()
...
        formatter->close_section(); // total 
        formatter->close_section(); // user
-     } 
-     formatter->dump_int("TotalBytes", header.stats.total_bytes);
-     formatter->dump_int("TotalBytesRounded", header.stats.total_bytes_rounded);
-     formatter->dump_int("TotalEntries", header.stats.total_entries);
+     }
+
+     if (s->cct->_conf->rgw_getusage_dump_description) {
+       formatter->open_object_section("Stats");
```
Comment (yehudasa) [inline src/rgw/rgw_rest_s3.cc]: @weiqiaomiao close if here, and reopen it for the close_section(), so that we don't duplicate the stats dump

**Comment by weiqiaomiao** [comment]
@yehudasa  Thanks for your suggestion, and the commit has been updated. Would you take a look at it again?

**Comment by yehudasa** [comment]
@weiqiaomiao this looks ok. Last thing we need is that since this affects compatibility, we need to note it in the relase notes. Please add a description of this change, the new configurable (to override), and the api change to ceph/PendingReleaseNotes. Thanks.

**Comment by weiqiaomiao** [comment]
@yehudasa  I already add this change to ceph/PendingReleaseNote. Would you take a look at it again.Thanks.

File: PendingReleaseNotes [hunk 3]
```diff
@@ -17,6 +17,24 @@
   listed in 'osd class default list' requires a capability naming the class
   (e.g. 'allow class foo').
 
+* The 'rgw_rest_getusage_op_compat' config option allow you to dump the description of user stats
```
Comment (theanalyst) [inline PendingReleaseNotes]: config option `allows`

File: PendingReleaseNotes [hunk 4]
```diff
@@ -17,6 +17,24 @@
...
+  or not in s3 GetUsage API.This config is bool type and its default value is false.
+  If the value is true, the reponse data of s3 GetUsage like bellow:
```
Comment (theanalyst) [inline PendingReleaseNotes]: `below`

File: PendingReleaseNotes [hunk 5]
```diff
@@ -17,6 +17,24 @@
...
+
+    "stats": {
+                "TotalBytes": 516,
+                "TotalBytesRounded": 1024,
+                "TotalEntries": 1
+            }
+
+  Or if the value is false, the  reponse of s3 GetUsage like bellow and as it do before we add this config:
```
Comment (theanalyst) [inline PendingReleaseNotes]: s/bellow/below,  and `as it do` -> `as it does`

Comment (theanalyst) [inline PendingReleaseNotes]: also preferably, `before we add this config option`

File: PendingReleaseNotes [hunk 6]
```diff
@@ -17,6 +17,24 @@
   listed in 'osd class default list' requires a capability naming the class
   (e.g. 'allow class foo').
 
+* The 'rgw_rest_getusage_op_compat' config option allow you to dump the description of user stats
+  or not in s3 GetUsage API.This config is bool type and its default value is false.
```
Comment (theanalyst) [inline PendingReleaseNotes]: space after period, and `this config is` -> `this config option is of` , also `its default value is false` -> `defaults to false`

**Comment by weiqiaomiao** [comment]
@theanalyst the commit has been updated according to your suggestion. would you take a look at it again. thanks!

File: PendingReleaseNotes (hunk 3 above)
Comment (yehudasa) [inline PendingReleaseNotes]: @weiqiaomiao also, no underscores needed when specifying the config option

**Comment by yehudasa** [comment]
@weiqiaomiao please note the two comments on the last commit

**Comment by weiqiaomiao** [comment]
@yehudasa Thank you for reminding me. The commit has been updated.

**Comment by yehudasa** [comment]
Created and merged another PR due to conflict, see https://github.com/ceph/ceph/pull/11325

**Comment by yehudasa** [comment]
@weiqiaomiao also updated commit message, noted tracker issue

**Pull request closed without merging**